### PR TITLE
Graph Optimizations Doc

### PR DESCRIPTION
**Pull request by askhade**
**Description**: Adding documentation and usage guidelines for graph optimizers.

### CODE REVIEW

File: docs/ONNX_Runtime_Graph_Optimizations.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+## Usage
+
+### General Note
+Levels: 
+ONNX Runtime defines GraphOptimizationLevel enum for these levels. Each level when enabled, enables all the preceeding levels too. Mapping of these levels to enum is as follows:
+
+GraphOptimizationLevel::ORT_DISABLE_ALL -> Disables all optimizations
+GraphOptimizationLevel::ORT_ENABLE_BASIC -> Enables basic optimizations
+GraphOptimizationLevel::ORT_ENABLE_EXTENDED -> Enables basic and extended optimizations
+GraphOptimizationLevel::ORT_ENABLE_ALL -> Enables all avialable optimizations including layout optimizations
```
Comment (askhade) [inline docs/ONNX_Runtime_Graph_Optimizations.md]: This does not show as bullet... will make the change for better visualization

File: docs/ONNX_Runtime_Graph_Optimizations.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+These optimizations change the data layout for applicable nodes to achieve higher performance improvements. They are run after graph partitioning and are only applied to nodes assigned to CPU execution provider. Available layout optimizations are as follows:
+
+* NCHWc Optimizer : Optimizes the graph by using NCHWc layout instead of NCHW layout.
+
+## Usage
+
+### General Note
+Levels: 
+ONNX Runtime defines GraphOptimizationLevel enum for these levels. Each level when enabled, enables all the preceeding levels too. Mapping of these levels to enum is as follows:
```
Comment (faxu) [inline docs/ONNX_Runtime_Graph_Optimizations.md]: Consider rewording: "Enabling each level also enables all preceding levels. For example, enabling level 3 also enables levels 2 and 1."

File: docs/ONNX_Runtime_Graph_Optimizations.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+* Extended
+* Layout Optimizations
+
+### Basic Graph Optimizations
+
+These are semantics preserving graph rewrites which remove redundant nodes and redundant computations. These optimizations are enabled by default. They run before graph partitioning and thus apply to all the execution providers. Available basic graph optimizations are as follows:
+
+* Constant Folding : Statically computes parts of the graph that rely only on constant initializers. This eliminates the need to compute them during runtime.
+
+* Redundant node eliminations : Removes all redeundant nodes without changing the graph structure
```
Comment (faxu) [inline docs/ONNX_Runtime_Graph_Optimizations.md]: typo on redundant

File: docs/ONNX_Runtime_Graph_Optimizations.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
+# Graph Optimizations in ONNX Runtime
+
+ONNX Runtime provides various graph level optimizations to improve model performance. Graph optimizations are essentially graph level rewrites ranging from small graph simplifications and node eliminations to more complex node fusions and layout optimizations. 
+
+Graph optimizations are divided in several levels\categories based on their complexity and function. They can be performed both online and offline. ONNX Runtime provides Python, C#, C++ and C apis to enable different optimization levels and to choose the offline vs online mode. In offline mode the runtime saves the optimized graph to the disk. **Note**: When running in offline mode make sure to use the exact same options and hardware as the target machine.
```
Comment (faxu) [inline docs/ONNX_Runtime_Graph_Optimizations.md]: capitalize APIs

File: docs/ONNX_Runtime_Graph_Optimizations.md (hunk 1 above)
Comment (faxu) [inline docs/ONNX_Runtime_Graph_Optimizations.md]: these lines run into each other. may want to bullet these. should the enum value also be listed?

File: docs/ONNX_Runtime_Graph_Optimizations.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+Levels: 
+ONNX Runtime defines GraphOptimizationLevel enum for these levels. Each level when enabled, enables all the preceeding levels too. Mapping of these levels to enum is as follows:
+
+GraphOptimizationLevel::ORT_DISABLE_ALL -> Disables all optimizations
+GraphOptimizationLevel::ORT_ENABLE_BASIC -> Enables basic optimizations
+GraphOptimizationLevel::ORT_ENABLE_EXTENDED -> Enables basic and extended optimizations
+GraphOptimizationLevel::ORT_ENABLE_ALL -> Enables all avialable optimizations including layout optimizations
+
+Modes:
+All these optimizations can be run offline too. For production scenarios where model startup time is critical this can bring a lot of benefit. In offline mode, after running graph optimizations onnx runtime serializes the resulting model to the disk. **Note**: When running in offline mode make sure to use the exact same options(execution providers, optimization level etc...) and hardware as the target machine.
```
Comment (faxu) [inline docs/ONNX_Runtime_Graph_Optimizations.md]: please capitalize ONNX Runtime as a proper noun.

**Comment by ybrnathan** [comment]
Could you please change the onnxruntime_c_api.h and replace below comments with a link to this doc?
TODO (askhade) Add documentation about which optimizations are enabled for each value.

**Pull request merged**